src/fixedrange.jl

```diff
@@ -0,0 +1,154 @@
+    xrange = xmax .- xmin
+    scale = upper - lower
+    nObs, nVars = size(X)
+
+    for iVar in 1:nVars
+        @inbounds for iObs in 1:nObs
+            X[iObs, iVar] = lower + (X[iObs, iVar] - xmin[iVar]) / xrange[iVar] * scale
+        end
+    end
+    lower, upper, xmin, xmax
+end
+
+function fixedrange!(X::AbstractMatrix, lower, upper, xmin, xmax, ::ObsDim.Constant{2})
+    xrange = xmax .- xmin
+    scale = upper - lower
+    nVars, nObs = size(X)
+
+    for iObs in 1:nObs
+        @inbounds for iVar in 1:nVars
+            X[iVar, iObs] = lower + (X[iVar, iObs] - xmin[iVar]) / xrange[iVar] * scale
+        end
+    end
+    lower, upper, xmin, xmax
+end
+
+function fixedrange!{N}(X::AbstractVector, lower::Real, upper::Real, xmin::Vector, xmax::Vector, ::ObsDim.Constant{N})
+    @assert length(xmin) == length(xmax) == length(X) 
+    xrange = xmax .- xmin
+    scale = upper - lower
+    nVars = length(X)
+
+    @inbounds for iVar in eachindex(X) 
+        X[iVar] = lower + (X[iVar] - xmin[iVar]) / xrange[iVar] * scale
+    end
+    lower, upper, xmin, xmax
+end
+
+function fixedrange!{N}(X::AbstractVector, lower::Real, upper::Real, xmin::Real, xmax::Real, ::ObsDim.Constant{N})
+    xrange = xmax - xmin
+    scale = upper - lower
+    nVars = length(X)
+
+    @inbounds for iVar in eachindex(X) 
+        X[iVar] = lower + (X[iVar] - xmin) / xrange * scale
+    end
+    lower, upper, xmin, xmax
+end
+
+
+immutable FixedRangeScaler 
+    lower::Float64
+    upper::Float64
+    xmin::Vector
+    xmax::Vector
+    obsdim::ObsDim.Constant{}
+
+    function FixedRangeScaler(lower, upper, xmin, xmax, obsdim)
+        @assert length(xmin) == length(xmax) 
+        new(lower, upper, xmin, xmax, convert(ObsDimension, obsdim))
+    end
+end
+
+
+function FixedRangeScaler{T<:Real}(X::AbstractArray{T}; obsdim=LearnBase.default_obsdim(X))
+    FixedRangeScaler(X, convert(ObsDimension, obsdim))
+end
+
+function FixedRangeScaler{T<:Real,M}(X::AbstractArray{T,M}, ::ObsDim.Last)
+    FixedRangeScaler(X, ObsDim.Constant{M}())
+end
+
+function FixedRangeScaler{T<:Real,M}(X::AbstractArray{T}, obsdim::ObsDim.Constant{M})
+    FixedRangeScaler(0, 1, vec(minimum(X, M)), vec(maximum(X, M)), obsdim)
+end
+
+function FixedRangeScaler{T<:Real}(X::AbstractArray{T}, lower, upper; obsdim=LearnBase.default_obsdim(X))
+    FixedRangeScaler(X, lower, upper, convert(ObsDimension, obsdim))
+end
+
+function FixedRangeScaler{T<:Real,M}(X::AbstractMatrix{T}, lower, upper, obsdim::ObsDim.Constant{M})
+    FixedRangeScaler(lower, upper, vec(minimum(X, M)), vec(maximum(X, M)), obsdim)
+end
+
+function FixedRangeScaler{T<:Real,M}(X::AbstractArray{T,M}, lower, upper, ::ObsDim.Last)
+    FixedRangeScaler(X, lower, upper, ObsDim.Constant{M}())
+end
+
+function StatsBase.fit{T<:Real}(::Type{FixedRangeScaler}, X::AbstractArray{T}; obsdim=LearnBase.default_obsdim(X))
+    FixedRangeScaler(X, obsdim=obsdim)
```

Comment:
If you use `FixedRangeScaler(X, convert(ObsDimension, obsdim))` here the function will be type stable (without any downside I could think of). In general a good rule of thumb is to never call a keyword based method internally if it can be avoided. They are just for convenience for the end user